### ISSUE TITLE
glibc (ldd) update causes dynamic dependency failure

### ISSUE THREAD

**Issue by hdclark** [linuxdeploy/linuxdeploy#210]
Glibc was recently updated to version 2.36. Using linuxdeploy on Arch Linux causes `ldd` output to report `linux-vdso.so.1` and the interpreter as dependencies, which break AppImage packaging.

Old `ldd` output:
```
$> ldd some_binary
        linux-vdso.so.1 (0x00007ffc9e5d7000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f2e8b17b000)
        libygor.so => /usr/lib/x86_64-linux-gnu/libygor.so (0x00007f2e8aca9000)
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f2e8aa9e000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f2e8aa84000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f2e8a8ab000)
        libgsl.so.27 => /usr/lib/x86_64-linux-gnu/libgsl.so.27 (0x00007f2e8a5cf000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f2e8a48b000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f2e8a468000)
        libgslcblas.so.0 => /usr/lib/x86_64-linux-gnu/libgslcblas.so.0 (0x00007f2e8a427000)
```

New `ldd` output (notice first two lines):
```
$> ldd some_binary
        linux-vdso.so.1 => linux-vdso.so.1 (0x00007ffca0cea000)
        /lib64/ld-linux-x86-64.so.2 => /usr/lib64/ld-linux-x86-64.so.2 (0x00007f8173466000)
        libygor.so => /usr/lib/libygor.so (0x00007f8172fa8000)
        libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x00007f8172d71000)
        libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x00007f8172d51000)
        libc.so.6 => /usr/lib/libc.so.6 (0x00007f8172b6d000)
        libgsl.so.27 => /usr/lib/libgsl.so.27 (0x00007f8172895000)
        libgslcblas.so.0 => /usr/lib/libgslcblas.so.0 (0x00007f8172853000)
        libm.so.6 => /usr/lib/libm.so.6 (0x00007f817276b000)
```

Attempting to package a binary results in the following linuxdeploy error:
```
            ...
            -- Deploying dependencies for existing files in AppDir --
            Deploying dependencies for ELF file ./AppDir/usr/bin/some_binary
            ERROR: Cannot deploy non-existing library file: /wip/linux-vdso.so.1
            ERROR: Failed to deploy dependencies for existing files
```

I'm not certain, but I suspect the problem is that `ldd` output is being filtered via regex at https://github.com/linuxdeploy/linuxdeploy/blob/master/src/core/elf_file.cpp#L217 , and the new `ldd` output has begun emitting `=>` separators for linux-vdso.so.1 and the interpreter.

**Comment by hdclark** [comment]
Here is an example of a daily CI build failing due to the glibc update on Arch: https://gitlab.com/hdeanclark/DICOMautomaton/-/pipelines/603600818

**Comment by TheAssassin** [comment]
Should probably be filtered, yes. Would you like to send a PR?

For the record, Arch is a maximally unsuitable environment for building AppImages, not just because it's not a "stable target" (it's pretty unpredictable what to find on rolling release distros in general and especially on Arch) but also because it's a rolling release distro and therefore doesn't produce any realistically usable bundles. So generally, Arch is not supported by linuxdeploy. In this case, obviously, Arch is not to blame, so this is a totally justified bug report. I could imagine that eventually this will tricke down into any distro.

**Comment by hdclark** [comment]
Indeed. I use Arch builds as a canary for upstream changes like this.

Please see https://github.com/linuxdeploy/linuxdeploy/pull/211 . I'm not currently able to test this code.

**Comment by hdclark** [comment]
Note: in case anyone needs a (terribly, hacky) workaround:

Step 1: replace system ldd script with a wrapper:
```
    mv /usr/sbin/ldd{,_orig}
    printf '#!/usr/bin/env bash\nldd_orig "$@" | grep -v linux-vdso | grep -v ld-linux\n' > /usr/sbin/ldd
    chmod 777 /usr/sbin/ldd
```

Step 2: run linuxdeploy:
```
    ./linuxdeploy-${ARCH}.AppImage ...
```

Step 3: reset system ldd:
```
    mv /usr/sbin/ldd{_orig,}
```

Alternatively, a shell alias might be safer...

**Comment by TheAssassin** [comment]
> I use Arch builds as a canary for upstream changes like this.

Never thought of this, but it's a really good idea! We should document that somewhere.

**Comment by kkarsten62** [comment]
Same problem here. I am using linuxdeploy within a dockerfile based on latest Manjaro "FROM manjarolinux/base:20220807".
@hdclark : Sorry, I can not understand your doing for the workaround. Can give me a helping hand?

I have the following in the docker file:
```
...
# Install AppImage
# See https://docs.appimage.org/packaging-guide/from-source/linuxdeploy-user-guide.html
RUN wget https://github.com/linuxdeploy/linuxdeploy/releases/download/continuous/linuxdeploy-x86_64.AppImage \
	&& wget https://github.com/linuxdeploy/linuxdeploy-plugin-qt/releases/download/continuous/linuxdeploy-plugin-qt-x86_64.AppImage \
	&& chmod +x linuxdeploy*.AppImage
...
DEPLOY_CMD="./linuxdeploy-x86_64.AppImage \
	--desktop-file /AppDir/usr/share/applications/myapp.desktop \
	--icon-file /AppDir/usr/share/icons/hicolor/128x128/apps/myapp.png \
	--appdir AppDir \
	--plugin qt \
	--custom-apprun=/apprun.sh \
	--output appimage"
$DEPLOY_CMD
...

```
Many thanks.
kkarsten62

**Comment by kkarsten62** [comment]
I have successfully integrate the workaround into my docker/linuxdeploy environment. AppImage for Manjaro based on glibc 2.36 will be build now. Many thanks to @hdclark for the workaround and the PR.

**Comment by hdclark** [comment]
I'm glad it worked for you. Here is my current workaround inside a bash script:

```
    ldd_path="$(which ldd)"
    printf 'Wrapping system ldd...\n'
    mv "${ldd_path}"{,_orig}
    printf '#!/usr/bin/env bash\nldd_orig "$@" | grep -v linux-vdso | grep -v ld-linux\n' > "${ldd_path}"
    chmod 777 "${ldd_path}"

    trap "printf 'Returning system ldd...\n' ; mv "${ldd_path}"{_orig,} ; exit ;" EXIT

    # run linuxdeploy / ldd and then exit when done. The original ldd should be returned when script exits.
```

Please be careful since the system ldd disappearing could cause problems!

**Comment by Narcha** [comment]
> If this library is not existing it's probably not used either. If linuxdeploy had an option to filter unused dependencies via `ldd --unused` this couild be a considerable workaround which probably also reduces some overhead.

This library is somewhat of a special case. It might be used, but it does not have any physical file on disk.
See https://stackoverflow.com/a/58657078/13118494

**Comment by Diyou** [comment]
Just tested ... ```ldd --unused``` does indeed not filter this particular library. I had this problem yesterday with 2 libraries missing from the system of which ```linux-vdso.so.1```was one and the other was not used.

**Comment by Diyou** [comment]
This [post](https://forum.endeavouros.com/t/glibc-2-36-gives-a-different-result-ldd-result-on-libvdso-so-1/30028) points out that the ldd entry is pointing on itself:
```linux-vdso.so.1 => linux-vdso.so.1 (0x00007ffca0cea000)```
I don't have a lot of experience with ldd but this might be a clue.

**Comment by Narcha** [comment]
@hdclark This can be closed, right?

**Comment by TheAssassin** [comment]
@Narcha apparently this is working as intended, so let's close the issue as resolved by 287c4364e2bc41b8a52304a5382253088c0342ac. In the unlikely case the issue persists, please reopen or comment here.
